File: equivalent/gateway/table.py
"""The precondition table: one row per action the gateway knows about.

Trust role: this is what POST /run checks before it will dispatch to
anything. A row with the wrong `requires` lets a check run on evidence
it should not trust, or blocks one that is actually ready.

Every row belongs to one phase. A region is in one phase for its whole
life, so the rows a session ever sees are the rows of its region's
phase; the table holds both because one gateway serves regions of both
kinds, and because a reader of an old ledger has to be able to look up
an action of either.
"""
from __future__ import annotations

from dataclasses import dataclass

from equivalent.ledger.acceptance import (
    ACCEPTANCE_REQUIREMENTS,
    ONBOARDING,
    ONBOARDING_REQUIREMENTS,
    PORTING,
    acceptance_requirements,
)

# The row that names a phase's whole requirement list rather than an
# action to dispatch. The porting one is the only row whose preconditions
# depend on the code, so it is spelled here for `requires_for` to find.
ACCEPT = "accept"

# What a config key means and what shape its value has, written once.
# GET /table serves this beside the row that declares the key, so the pi
# extension can offer the key as a typed tool parameter without keeping a
# second copy of the wording, and POST /run checks a value against the
# same type before it hashes the config.
CONFIG_KEY_SPECS = {
    "repeats": {
        "type": "integer",
        "description": "How many timed runs to make. Five when left out.",
    },
    "seed": {
        "type": "integer",
        "description": (
            "The Hypothesis seed for the search. One is drawn for the run "
            "when left out, and the claim records which."
        ),
    },
    "max_examples": {
        "type": "integer",
        "description": "How many examples to draw per property. A hundred when left out.",
    },
    "limit": {
        "type": "integer",
        "description": "Score at most this many mutants, rather than all of them.",
    },
}


@dataclass(frozen=True)
class ActionRow:
    name: str
    emits: tuple  # tuple[str, ...] -- predicate types this action can produce
    requires: tuple  # tuple[tuple[str, str], ...] -- (predicate_type, subject_kind) pairs
    deterministic: bool
    # None only for the row that names a phase's whole requirement list --
    # "accept" and "onboarded" -- which has nothing to dispatch to.
    component: str | None
    # Which kind of session this action belongs to. It is written on every
    # row rather than inferred from the name so that GET /table can serve
    # one phase's rows without a second list saying which those are.
    phase: str
    # The config keys POST /run accepts for this action; anything else is
    # rejected before dispatch. This keeps the config canonical, so
    # duplicate detection (which hashes the config) can't be defeated by
    # padding a request with junk keys until it hashes differently. GET
    # /table serves these keys with the row, so a client can offer them
    # without knowing the table; every one of them needs an entry in
    # CONFIG_KEY_SPECS above to say what it is.
    config_keys: tuple = ()


ACTION_TABLE = (
    # Porting: one region of a code that has already been brought in.
    ActionRow("sese_check", ("sese/verified",), (), True, "analyzer:check_sese", PORTING),
    ActionRow("build_replay", ("build/replay",), (("sese/verified", "frozen"),), True,
              "builder:/v1/build", PORTING),
    ActionRow("run_replay", ("gpu/executed",), (("build/replay", "tree"),), True,
              "builder:/v1/run", PORTING),
    ActionRow(
        "sanitize", ("sanitize/memcheck", "sanitize/racecheck", "sanitize/initcheck"),
        (("gpu/executed", "tree"),), True, "builder:/v1/sanitize", PORTING,
    ),
    ActionRow(
        "regression_visible", ("regression/visible",),
        (("sanitize/memcheck", "tree"), ("sanitize/racecheck", "tree")), True,
        "oracle:/v1/compare", PORTING,
    ),
    ActionRow(
        "property_check", ("regression/property",),
        (("regression/visible", "tree"),), True, "builder:/v1/properties", PORTING,
        config_keys=("seed", "max_examples"),
    ),
    ActionRow(
        "regression_holdout", ("regression/holdout",),
        (("regression/visible", "tree"),), True, "oracle:/v1/compare", PORTING,
    ),
    ActionRow(
        "program_regression", ("program/regression",),
        (("regression/holdout", "tree"),), True, "builder:/v1/time", PORTING,
    ),
    ActionRow("time_port", ("timing/port",), (("program/regression", "tree"),), False,
              "builder:/v1/time", PORTING, config_keys=("repeats",)),
    ActionRow("time_baseline", ("timing/baseline",), (), False, "builder:/v1/time", PORTING,
              config_keys=("repeats",)),
    ActionRow(
        "accept", (), tuple((r.predicate_type, r.subject_kind) for r in ACCEPTANCE_REQUIREMENTS),
        True, None, PORTING,
    ),

    # Onboarding: bringing a code in, in the order one step's evidence
    # becomes the next step's precondition.
    ActionRow("manifest_check", ("manifest/valid",), (), True, "gateway:manifest_check", ONBOARDING),
    ActionRow("harness_build", ("harness/builds",), (("manifest/valid", "tree"),), True,
              "builder:/v1/build", ONBOARDING),
    ActionRow("harness_capture", ("harness/captured",), (("harness/builds", "tree"),), True,
              "builder:/v1/capture", ONBOARDING),
    ActionRow("harness_replay", ("harness/replays",), (("harness/captured", "tree"),), True,
              "builder:/v1/run", ONBOARDING),
    ActionRow("harness_determinism", ("harness/deterministic",), (("harness/replays", "tree"),),
              True, "builder:/v1/capture", ONBOARDING),
    ActionRow("harness_timing", ("harness/times",), (("harness/builds", "tree"),), True,
              "builder:/v1/time", ONBOARDING),
    ActionRow("harness_self_check", ("harness/self_check",), (("harness/replays", "tree"),),
              True, "builder:/v1/mutate", ONBOARDING, config_keys=("limit",)),
    ActionRow("harness_property", ("harness/properties",), (("harness/replays", "tree"),),
              True, "builder:/v1/properties", ONBOARDING,
              config_keys=("seed", "max_examples")),
    ActionRow(
        "onboarded", (), tuple((r.predicate_type, r.subject_kind) for r in ONBOARDING_REQUIREMENTS),
        True, None, ONBOARDING,
    ),
)


def rows_for(phase: str) -> tuple:
    """The rows a region in this phase may run, in the order they are listed."""
    return tuple(row for row in ACTION_TABLE if row.phase == phase)


def requires_for(row: ActionRow, manifest=None) -> tuple:
    """One row's preconditions for a particular code.

    Every row but one has the preconditions written on it. The `accept`
    row is the exception: what finishes a port depends on whether the code
    declares a module of invariants, so its list is read from the code's
    manifest rather than frozen into the table.
    """
    if row.name != ACCEPT:
        return row.requires
    return tuple(
        (req.predicate_type, req.subject_kind) for req in acceptance_requirements(manifest)
    )


def config_params(row: ActionRow) -> dict:
    """The type and wording of every config key this row accepts."""
    return {key: dict(CONFIG_KEY_SPECS[key]) for key in row.config_keys}

File: pi-extension/src/table.ts
/**
 * Trust role: none. This only turns the gateway's own precondition table
 * into tool descriptions; it decides nothing. A wrong description makes
 * the model call things in the wrong order, and the gateway's refusal is
 * what actually stops a premature call.
 */

import { Type, type TObject } from "typebox";

/** What one setting is worth: the gateway sends the wording, not this file. */
export interface ConfigParam {
  type: string;
  description: string;
}

export interface ActionRow {
  name: string;
  emits: string[];
  requires: [string, string][];
  deterministic: boolean;
  component: string | null;
  // The settings POST /run takes for this action, and what each one
  // means. Optional here because a row served by a gateway that predates
  // them arrives without them, in which case the action takes none.
  config_keys?: string[];
  config_params?: Record<string, ConfigParam>;
}

const SUBJECT_PHRASE: Record<string, string> = {
  tree: "the submitted tree",
  frozen: "this region",
  capture_set: "the capture set",
  strategy: "the strategy",
  binary: "the binary",
  outputs: "the outputs",
};

function subjectPhrase(kind: string): string {
  return SUBJECT_PHRASE[kind] ?? kind;
}

export function requiresLine(row: ActionRow): string | undefined {
  if (row.requires.length === 0) return undefined;
  const parts = row.requires.map(
    ([predicateType, subjectKind]) => `${predicateType} = pass for ${subjectPhrase(subjectKind)}`,
  );
  return `Requires: ${parts.join("; ")}.`;
}

/**
 * The settings this action takes, keyed as POST /run wants them. A key
 * the gateway declared but did not describe is left out: without a type
 * there is nothing to offer the model, and the gateway would reject a
 * guess anyway.
 */
export function configParams(row: ActionRow): Record<string, ConfigParam> {
  const described = row.config_params ?? {};
  const params: Record<string, ConfigParam> = {};
  for (const key of row.config_keys ?? []) {
    const spec = described[key];
    if (spec) params[key] = spec;
  }
  return params;
}

export function settingsLine(row: ActionRow): string | undefined {
  const entries = Object.entries(configParams(row));
  if (entries.length === 0) return undefined;
  const parts = entries.map(([key, spec]) => `${key} (${spec.type}) -- ${spec.description}`);
  return `Optional settings: ${parts.join(" ")}`;
}

export function describeRow(row: ActionRow): string {
  const produces = row.emits.length > 0 ? `Records: ${row.emits.join(", ")}.` : "Records no claim.";
  const lines = [`Runs '${row.name}' via ${row.component}.`, produces];
  const req = requiresLine(row);
  if (req) lines.push(req);
  const settings = settingsLine(row);
  if (settings) lines.push(settings);
  return lines.join(" ");
}

/**
 * The tool's parameters are the row's own settings: whatever the gateway
 * says this action takes, offered with the gateway's own wording, so the
 * two cannot drift. An action that takes none gets an empty object, as
 * before.
 */
export function toolParameters(row: ActionRow): TObject {
  const properties: Record<string, ReturnType<typeof Type.Optional>> = {};
  for (const [key, spec] of Object.entries(configParams(row))) {
    if (spec.type !== "integer") continue;
    properties[key] = Type.Optional(Type.Integer({ description: spec.description }));
  }
  return Type.Object(properties);
}

/**
 * What to send as the run's config: the declared settings the model
 * filled in, and nothing else. The gateway hashes this config to spot a
 * repeat of the same request, so anything it did not declare has no
 * business in it.
 */
export function runConfig(row: ActionRow, params: unknown): Record<string, unknown> {
  const given = (params ?? {}) as Record<string, unknown>;
  const config: Record<string, unknown> = {};
  for (const key of Object.keys(configParams(row))) {
    if (given[key] !== undefined) config[key] = given[key];
  }
  return config;
}

/**
 * Rows with no component (only "accept") have nothing for POST /run to
 * dispatch to -- the gateway 400s rather than refusing normally. GET
 * /status already reports acceptance, so these rows are excluded here
 * rather than given a tool that always errors.
 */
export function callableRows(table: ActionRow[]): ActionRow[] {
  return table.filter((row) => row.component !== null);
}
